File: leetcode/Algorithms/133.CloneGraph/Solution.py
class Solution:
    def __init__(self):
        self.map = {}

    def cloneGraph(self, node):
        if node is None:
            return None

        next = UndirectedGraphNode(node.label)
        self.map[str(node.label)] = next
        for tmp in node.neighbors:
            if str(tmp.label) in self.map:
                next.neighbors.append(self.map.get(str(tmp.label)))
            else:
                next.neighbors.append(self.cloneGraph(tmp))
        return self.map.get(str(node.label))

File: leetcode/Algorithms/133.CloneGraph/Solution.java
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Queue;

public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if (node == null) return null;

        HashMap<Integer, UndirectedGraphNode> map = new HashMap<>();
        map.put(node.label, new UndirectedGraphNode(node.label));

        Queue<UndirectedGraphNode> queue = new LinkedList<>();
        queue.add(node);

        while (queue.size() != 0) {
            UndirectedGraphNode origin = queue.poll(), next = map.get(origin.label);
            for (UndirectedGraphNode tmp : origin.neighbors) {
                if (!map.containsKey(tmp.label)) {
                    map.put(tmp.label, new UndirectedGraphNode(tmp.label));
                    queue.add(tmp);
                }
                next.neighbors.add(map.get(tmp.label));
            }
        }

        return map.get(node.label);
    }
}
